### PR TITLE
[IBC] Implement ICS-24 - Tracking IBC store transitions in the network state

### LINKED ISSUE #840: [IBC] Track and Include the IBC store in the network state (ICS-24)
## Objective

Add a new state tree that tracks the IBC store state for the entire chain.

Implement IBC message types that fulfil the `Message` interface defined in utility. These messages should define the update of key-value pairs in the IBC stores or the removal of a key from the local IBC store.

These messages should be propagated through the network and upon their receipt each node should add them to the transactions mempool managed by utility in order for them to be included in a new block and ultimately change the IBC state tree. This should include the logic to add any changes to a postgres table and from this set of updates update the IBC store state tree to include these changes.

## Origin Document

[IBC Store Outline](https://hackmd.io/@8Qg_Y9pwREeXqD4kIK5U_Q/HJyKTr9vh)

## Goals

- Support the tracking and inclusion of IBC related transition events in the Pocket blockchain by leveraging the existing message handler design in the utility module
  - Add IBC store state tree to the `TreeStore`
  - Create an IBC message type and handler method in the IBC module to propogate the changes to the IBC store
  - Implement the utility `Message` interface for the IBC message types for their use in the Transactions Mempool
  - Integrate the IBC message handling into the block production/application logic
- Implement a store manager to interact with the IBC store state tree locally to each host and send IBC messages on any changes

## Deliverable

- [ ] Add IBC store state tree
- [ ] Add IBC message types
- [ ] Add IBC message handler
- [ ] Add IBC message -> utility message conversion
  - [ ] Integrate mempool reaping with block production/application
- [ ] Add IBC store manager to interact with the stores locally and propagate IBC messages on changes

## Non-goals / Non-deliverables

- Change any of the existing block production/application logic

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **Task specific tests or benchmarks**: `make ...`
- [ ] **New tests or benchmarks**: `make ...`
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)
- [ ] **k8s LocalNet**: verify a `k8s LocalNet` is still functioning correctly by following the instructions [here](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md)

---

**Creator**: @h5law
**Co-Owners**: @Olshansk 

```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by h5law**
### Summary generated by Reviewpad on 12 Jul 23 20:45 UTC
This pull request includes the following changes:

- The file `main_test.go` was newly added to the repository and includes various changes such as package definition, imports, variable declaration, and function definitions related to IBC modules testing.

- The file `proofs_ics23.go` includes changes in error handling for the functions `createMembershipProof` and `createNonMembershipProof`, replacing the error `coreTypes.ErrCreatingProof` with `coreTypes.ErrIBCCreatingProof` for more specific error messages.

- The file `keys_ics03.go` was renamed from the `ibc/host` directory to the `ibc/path` directory. Additionally, the package declaration was changed from `host` to `path`. These changes seem related to the organization of the file within the codebase.

- The `config.validator2.json` file includes changes in the `ibc` section, adding two new keys: `private_key` and `stores_dir`. The `enabled` key remains unchanged. The `private_key` key has been assigned a long string value, and the `stores_dir` key has been assigned the value "/var/ibc".

- The file `internal/testutil/ibc/mock.go` includes renaming of the functions `BaseIbcMock` to `BaseIBCMock` and `IbcMockWithHost` to `IBCMockWithHost`. These changes update the function names to follow proper naming conventions.

- The file includes the implementation of a store manager for IBC (Inter-Blockchain Communication), responsible for managing provable stores, caching, pruning, and restoring data.

- The addition of a new function, `GetIBCStoreUpdates`, which retrieves the set of key-value pairs updated at the current height for the IBC store.

- Changes in the state hashes used in the tests in the file `state_test.go`.

- The file `runtime/manager_test.go` includes the addition of new fields `PrivateKey` and `StoresDir` to the `IBC` struct in the `NewManagerFromReaders` function.

- Changes in the `config.validator4.json` file include modifications in the `ibc` section, adding two new properties: `private_key` and `stores_dir`. The `enabled` property remains unchanged.

- The file `prefix.go` was renamed from the `ibc/host` package to the `ibc/path` package. The similarity index indicates that most of the content remains unchanged.

- There was a change in error handling logic in a file involving the renaming of the error variable and addressing the error during the creation of a new public key.

- The file `gov.go` includes changes in import statements, variable declarations, and function modifications related to new IBC store transaction fees.

- The file `persistence/types/ibc.go` is a new file added to the codebase, containing functions and constants related to IBC.

- The file `persistence/types/ibc.go` in the `persistence` package includes changes, introducing methods for setting and retrieving key-value pairs in the IBC store table.

- The file `persistence/trees.go` includes changes related to the addition of a new tree and corresponding functions to handle updates for the tree.

- The file `utils_test.go` includes changes in the casing of a function call from `ibcUtils.IbcMockWithHost` to `ibcUtils.IBCMockWithHost`.

- The file `ibc/host/path_test.go` was renamed to `ibc/path/path_test.go` with no other changes in the file.

- Changes in default variable values related to IBC functionality.

- The file `db.go` includes a new function, `initialiseIBCTables`, responsible for initializing an IBC store table in a PostgreSQL database.

- A new file, `ibc.feature.wip`, was added, containing integration and end-to-end tests for various scenarios related to IBC functionality.

- Changes in the `PROTOCOL_STATE_HASH.md` file include additions and modifications related to the IBC state tree and the handling of IBCMessage objects.

- The file `ibc/host.go` includes changes in comments, import statements, and the addition of fields and methods related to the host's interaction with the IBC store.

- A new file, `convert.go`, was added in the `ibc/types` directory, containing functions and imports to work with IBC messages.

- Changes in the `utils_test.go` file include modifications in the casing of a function call from `ibcUtils.IbcMockWithHost` to `ibcUtils.IBCMockWithHost`.

- The file `ibc/host/path_test.go` was renamed to `ibc/path/path_test.go` with no other changes in the file.

- Changes in default variable values related to IBC functionality.

- A new file, `db.go`, was added, containing a function

## Issue

Fixes #840 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Add new `IbcMessage` protobuf with 2 types
  - `UpdateIbcStore` and `PruneIbcStore`
- Track IBC store changes in the mempool
- Update IBC state tree using these state transition messages

## Testing

- [x] `make develop_test`; if any code changes were made
- [x] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [x] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [x] I have updated the corresponding README(s); local and/or global
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by h5law** [comment]
_Note_: This PR is based ontop of #843 and #842 (inturn also on #756) but the main logic for the mempool interactions, ICS-23 proof generation/verification and ICS-24 stores is in place

File: ibc/handle_message_test.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
+package ibc
```
Comment (h5law) [inline ibc/handle_message_test.go:1]: This is the testing of the message handling and would like some feedback on any other tests to go in here/expand on any of the current ones & any tests to go in persistence to test the Postgres side of things. Just high level what would be nice to see

@Olshansk @dylanlott

File: ibc/host.go [hunk 2]
```diff
@@ -1,8 +1,14 @@
+// TECHDEBT(#854): The host should be a submodule with access to the bus, this allows for the
+// host to access the persistence module and thus the tree store in order to create local
+// copies of the IBC state tree with the GetProvableStore() function. Bus access also will
+// allow for the host to send any local changes to these stores as IbcMessage types through
+// the P2P module's Broadcast() function to allow them to be included in the next block.
```
Comment (Olshansk) [inline ibc/host.go:5]:
```suggestion
// the P2P module's Broadcast() function to allow them to propagate through the network's mempool.
```

File: ibc/host.go [hunk 3]
```diff
@@ -1,8 +1,14 @@
+// TECHDEBT(#854): The host should be a submodule with access to the bus, this allows for the
```
Comment (Olshansk) [inline ibc/host.go:1]: @dylanlott I believe this is why we need to register the tree submodule rather than just set the bus. So its accessible from here

File: ibc/handle_message_test.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	idxTx := prepareIndexedMessage(t, persistenceMod.GetTxIndexer())
+
+	// Error on having an indexed transaction
+	err := utilityMod.HandleTransaction(idxTx.Tx)
+	require.Error(t, err)
+	require.EqualError(t, err, coreTypes.ErrTransactionAlreadyCommitted().Error())
+}
+
+func TestHandleMessage_BasicValidation_Message(t *testing.T) {
+	updateMsg, _ := prepareUpdateMessage(t, []byte("key"), []byte("value"))
```
Comment (Olshansk) [inline ibc/handle_message_test.go:45]: For this tet and the one below, I don't fully understand why you have a "mini test" above the `testCases`. Can you add some comments?

File: ibc/handle_message_test.go (hunk 1 above)
Comment (Olshansk) [reply ibc/handle_message_test.go:1]: @h5law From a unit testing perspective, this is pretty extensive and you did a good job. 

That said, I understand where your concern / lack of confidence comes from and unfortunately we'll have to live with it until we've made some progress. I have two suggestions:

1. Pipe this whole `git diff` through ChatGPT and say 

> You are a senior engineer with experience in protocol development, and expertise in the IBC protocol. Review the tests I have implemented in `handle_message_test` and provide some additional test cases that I have not accounted for yet.

2. Though we don't have the infra / CLI to test this E2E, do the following:
2.1 Create a placeholder `ibc.feature` and write out (in plain English) what kind of tests you want to have in the future
2.2 Create a new ticket to implement them later and leave a TODO

File: ibc/host.go [hunk 5]
```diff
@@ -16,3 +22,11 @@ type host struct {
 func (h *host) GetTimestamp() uint64 {
 	return uint64(time.Now().Unix())
 }
+
+// GetProvableStore returns a copy of the IBC state tree where all operations happen
+// locally (in memory) and are not persisted to the database. All changes are instead
+// broadcasted to the network for inclusion in the next block.
+// TODO(#854): Implement this
```
Comment (Olshansk) [inline ibc/host.go:29]:
```suggestion
// GetProvableStore returns a copy of the IBC state tree where all operations observed by 
// this specific ibc light client were applied to its ephemeral (in memory) state and have not
// yet been included in the next block. The aggregation of all light client-provable stores
// propagated throughout the network are happen validated by the block proposer when reaping
// the mempool, and lead to a valid on-chain state transition when consensus is reached.
// TODO(#854): Implement this
```

**Review by Olshansk — CHANGES_REQUESTED** [review]
Reviewed a handful of files and replied to the question around testing.

Lmk if it helps and also please rebased this and the other branches with master when you have a chance!

File: runtime/configs/proto/ibc_config.proto [hunk 6]
```diff
@@ -6,4 +6,5 @@ option go_package = "github.com/pokt-network/pocket/runtime/configs";
 
 message IBCConfig {
     bool enabled = 1;
+    string private_key = 2; // hex encoded
```
[the remark this answers is not in the document]
Comment (h5law) [reply runtime/configs/proto/ibc_config.proto:9]: This is a private key used to sign the `IbcMessage` types once they get put into a transcation so they pass validation entering the mempool

File: ibc/docs/ics24.md [hunk 7]
```diff
@@ -27,6 +55,60 @@ Identifiers are bytestrings constrained to specific characters and lengths depen
 
 The `GetTimestamp()` function returns the current unix timestamp of the host machine and is used to calculate timeout periods for packets
 
+## IBC State
+
+As mentioned [above](#persistence) the IBC store **must** be included in the consensus state of the network. As such the IBC store as defined in [ICS-24][ics24] has been implemented as a single IBC state tree.
+
+### IBC State Tree
+
+The IBC state tree is a non-value hashing `SMT` backed by a persistent `KVStore`, this is due to its need for data retrieval as well as proof generation/verification. The root hash of the IBC state tree is included in the `rootTree` which computes the networks state hash for any given block. This allows verifiers to not only verify the inclusion/exclusion of any element in the IBC state tree itself but also that the IBC state tree was used to compute the network's state hash, by utilising the `CommitmentProof` object defined in [ICS-23][ics23].
```
[the remark this answers is not in the document]
Comment (h5law) [reply ibc/docs/ics24.md:64]: If we were to read from a SMT without using the `WithValueHasher(nil)` option in its creation we will **only** get the values hash. I don't see where the current state trees are being used for retrieval but I may be wrong. This PR makes all state trees non-value hashing. In a non value hashing tree when we do `tree.Get([]byte("key"))` we get the actual value, not its hash.

**Comment by h5law** [comment]
🚨 I REWROTE HISTORY 🚨

This PR was based off of numerous other branches which have now been merged into main. Because of this I have rebased entirely off of `ibc/ics23_integration` and this the commit history has changed.

File: persistence/types/ibc.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package types
+
+import (
+	"encoding/hex"
+	"fmt"
+)
+
+const (
+	IbcStoreTableName   = "ibc_messages"
```
Comment (Olshansk) [inline persistence/types/ibc.go:9]: `s/ibc_messages/ibc_entries`? Since they're key-value pairs?

File: persistence/trees/trees.go [hunk 9]
```diff
@@ -38,6 +43,7 @@ const (
 	TransactionsTreeName = "transactions"
 	ParamsTreeName       = "params"
 	FlagsTreeName        = "flags"
+	IbcTreeName          = "ibc"
```
Comment (Olshansk) [inline persistence/trees/trees.go:46]: NIT: `IBCTreeName`. Ditto elsewhere


Screenshot 2023-06-28 at 10 18 06 PM

**Review by Olshansk — CHANGES_REQUESTED** [review]
Reviewed most of the PR but still have to get into the meat of things

Screenshot 2023-06-28 at 10 19 52 PM


I think we should be able to get this in soon!

File: ibc/docs/ics24.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (h5law) [reply ibc/docs/ics24.md:103]: So the idea behind the ADR is very clear and makes me think its a good idea. This would be very good to flesh out on a call but the way we interact with the stores will have to change. The following is an idea of how this would work (the emit functions are found in the ICS-24 submodule PR yet to be put up)

```go
type StoreManager struct {
	stores map[string]*ProvableStore
}

func (s *StoreManager) PruneCaches(height uint64) error {
	for _, store := range s.stores {
		if err := store.PruneCache(height); err != nil {
			return err
		}
	}
	return nil
}

// CachedEntry represents a local change made to the IBC store prior to it being
// committed to the state tree. These should be written to disk in the to prevent a
// loss of data and pruned when included in the state tree
// written to disk as follows:
// "{height}/{prefixedKey}" => value
type CachedEntry struct {
	height uint64
	prefixedKey []byte
	value []byte
}

func (c *CachedEntry) prepare() ([]byte, []byte) {
	return []byte(fmt.Sprintf("%d/%s", c.height, string(c.prefixedKey))), value
}

// ProvableStore is a struct that interfaces with the PostgresDB instance
// obtained from Persistence. It is used to Get/Set/Delete the keys in the
// IBC state tree, in doing so it will trigger the creation of 
type ProvableStore struct {
	bus modules.Bus // used to interact with persistence (passed from IBCHost)
	name string
	storeDir string // /var/stores/{name}
	prefix coreTypes.CommitmentPrefix // []byte(name)
	cache []*CachedEntry
}

func (p *ProvableStore) Get(key []byte) ([]byte, error) {
	prefixed := applyPrefix(p.prefix, key)
	persistenceMod := p.bus.GetPersistenceModule()
	currHeight := p.bus.GetConsensusModule().CurrentHeight()
	return persistenceMod.GetIBCEntry(prefixed, int64(currHeight)-1) // returns latest value stored
}

func (p *ProvableStore) GetAndProve(key []byte, membership bool) ([]byte, *ics23.CommitmentProof, error) {
	prefixed := applyPrefix(p.prefix, key)
	persistenceMod := p.bus.GetPersistenceModule()
	currHeight := p.bus.GetConsensusModule().CurrentHeight()
	value, err := persistenceMod.GetIBCEntry(prefixed, int64(currHeight)-1) // returns latest value stored
	if err != nil {
		return nil, nil, err
	}
	// import IBC state tree
	root, nodeStore := p.bus.GetTreeStore()
	lazy := smt.ImportSparseMerkleTree(nodeStore, root, sha256.New())
	proof, err := lazy.Prove(prefixed)
	if err != nil {
		return nil, nil, err
	}
	// convert SMT proof into CommitmentProof
	ics23Proof := new(ics23.CommitmentProof)
	if membership {
		ics23Proof, err = createMembershipProof(proof, prefixed, value)
	} else {
		ics23Proof, err = createNonMembershipProof(proof, prefixed)
	}
	return value, ics23Proof, nil
}

func (p *ProvableStore) Set(key, value []byte) error {
	prefixed := applyPrefix(p.prefix, key)
	persistenceMod := p.bus.GetPersistenceModule()
	currHeight := p.bus.GetConsensusModule().CurrentHeight()
	if err := persistenceMod.SetIBCEntry(prefixed, value); err != nil {
		return err
	}
	return emitUpdateStoreEvent(p.prefix, key, value)
}


func (p *ProvableStore) Delete(key []byte) error {
	prefixed := applyPrefix(p.prefix, key)
	persistenceMod := p.bus.GetPersistenceModule()
	currHeight := p.bus.GetConsensusModule().CurrentHeight()
	if err := persistenceMod.SetIBCEntry(prefixed, nil); err != nil {
		return err
	}
	return emitPruneStoreEvent(p.prefix, key, value)
}

func (p *ProvableStore) CacheEntries() error {
	store, err := kvstore.NewKVStore(p.storeDir)
	if err != nil {
		return err
	}
	for _, entry := range p.cache {
		key, value := entry.prepare()
		if err := store.Set(key, value); err != nil {
			return err
		}
	}
	p.cache := make([]*CachedEntry, 0)
	return store.Stop()
}

func (p *ProvableStore) PruneCache(height uint64) error {
	store, err := kvstore.NewKVStore(p.storeDir)
	if err != nil {
		return err
	}
	keys, _, err := store.GetAll([]byte(fmt.Sprintf("%d", height)), false)
	if err != nil {
		return err
	}
	for _, key := range keys {
		if err := store.Delete(key); err != nil {
			return err
		}
	}
	return store.Stop()
}
```

File: shared/core/types/error.go [hunk 10]
```diff
@@ -317,9 +321,13 @@ const (
 	UnknownActorTypeError             = "the actor type is not recognized"
 	UnknownMessageTypeError           = "the message being by the utility message is not recognized"
 	HostAlreadyExistsError            = "an ibc host already exists"
+	HostDoesNotExistError             = "an ibc host does not exist"
```
[the remark this answers is not in the document]
Comment (h5law) [reply shared/core/types/error.go:324]: Done, but do you mean `HostDoesNotExistError` => `IBCHostDoesNotExistError` or in the error itself?

File: ibc/types/messages.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+	return string(m.ProtoReflect().Descriptor().Name())
+}
+
+func (m *UpdateIbcStore) GetMessageRecipient() string { return "" }
+func (m *PruneIbcStore) GetMessageRecipient() string  { return "" }
+
+func (m *UpdateIbcStore) GetActorType() coreTypes.ActorType {
+	return coreTypes.ActorType_ACTOR_TYPE_VAL
+}
+
```
[the remark this answers is not in the document]
Comment (h5law) [reply ibc/types/messages.go:63]: go fmt does not allow

File: ibc/docs/ics24.md [hunk 12]
```diff
@@ -1,22 +1,89 @@
...
+- [IBC State](#ibc-state)
+  - [IBC State Tree](#ibc-state-tree)
+  - [IBC Messages](#ibc-messages)
+  - [IBC Message Handling](#ibc-message-handling)
+  - [Mempool](#mempool)
+  - [State Transition](#state-transition)
 
 ## Overview
 
 [ICS-24][ics24] details the requirements of the host chain, in order for it to be compatible with IBC. A host is defined as a node on a chain that runs the IBC software. A host has the ability to create connections with counterparty chains, open channels, and ports as well as commit proofs to the consensus state of its own chain for the relayer to submit to another chain. The host is responsible to managing and creating clients and all other aspects of the IBC module.
```
Comment (dylanlott) [inline ibc/docs/ics24.md:18]:
```suggestion
[ICS-24][ics24] details the requirements of the host chain, in order for it to be compatible with IBC. A host is defined as a node on a chain that runs the IBC software. A host has the ability to create connections with counterparty chains, open channels, expose ports, and commit proofs to the consensus state of its own chain for the relayer to submit to another chain. The host is responsible to managing and creating clients and all other aspects of the IBC module.
```

File: ibc/docs/ics24.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (dylanlott) [reply ibc/docs/ics24.md:103]: @h5law def let's hop on a call today about this 👍

File: shared/core/types/error.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply shared/core/types/error.go]: In the variable name so what you did is perf

File: ibc/host.go (hunk 3 above)
Comment (dylanlott) [reply ibc/host.go:1]: For posterity: TreeStore registration on the bus is added by #861

File: ibc/store/provable_store.go [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+	currHeight := int64(p.bus.GetConsensusModule().CurrentHeight())
+	rwCtx, err := p.bus.GetPersistenceModule().NewRWContext(currHeight)
+	if err != nil {
+		return err
+	}
+	defer rwCtx.Release()
+	if err := rwCtx.SetIBCStoreEntry(prefixed, nil); err != nil {
+		return err
+	}
+	p.cache = append(p.cache, &cachedEntry{
```
Comment (Olshansk) [inline ibc/store/provable_store.go:162]: IMPORTANT COMMENT.

What if someone sends multiple `Store` and `Prune` operations for the same key throughout the duration of a single block?

I believe the behaviour (and expected outcome) is arguably unexpected. My suggestion is to update the `cache` to be a map. Consider even looking at `GenericFIFOSet` and `txFIFOMempool` of how its used.

The outcome is still non-deterministic (it is subdued to race conditions) but I think its a better approach

File: ibc/store/provable_store.go [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+func (p *provableStore) Root() ics23.CommitmentRoot {
+	// TODO(#854): Implement tree retrieval
+	/**
+	root, _ := p.bus.GetTreeStore().GetTree(trees.IBCStateTree)
+	return root
+	**/
+	return nil
+}
+
+// CacheEntries writes all local changes to disk and clears the in-memory cache
```
Comment (Olshansk) [inline ibc/store/provable_store.go:186]: s/CacheEntries/FlushEntries

Flush is the proper term for this since you're moving from cache to disk
https://stackoverflow.com/a/48331611/768439

File: persistence/trees/trees.go [hunk 15]
```diff
@@ -341,3 +350,23 @@ func (t *treeStore) updateFlagsTree(flags []*coreTypes.Flag) error {
 
 	return nil
 }
+
+func (t *treeStore) updateIBCTree(keys, values [][]byte) error {
+	if len(keys) != len(values) {
+		return fmt.Errorf("keys and values must be the same length")
+	}
+	for i := 0; i < len(keys); i++ {
```
Comment (Olshansk) [inline persistence/trees/trees.go:358]: for i, key := range keys {

File: persistence/trees/trees.go [hunk 16]
```diff
@@ -341,3 +350,23 @@ func (t *treeStore) updateFlagsTree(flags []*coreTypes.Flag) error {
...
+		value := values[i]
+		if value == nil {
+			if err := t.merkleTrees[IBCTreeName].tree.Delete(key); err != nil {
+				return err
+			}
+		} else {
+			if err := t.merkleTrees[IBCTreeName].tree.Update(key, value); err != nil {
+				return err
+			}
+		}
```
Comment (Olshansk) [inline persistence/trees/trees.go:369]: i just think it's easier on the eys

```suggestion
		var err error
		if value == nil {
			err = t.merkleTrees[IBCTreeName].tree.Delete(key)
		} else {
			err := t.merkleTrees[IBCTreeName].tree.Update(key, value)
		}
		return err
```

File: ibc/store/provable_store.go [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+			return err
+		}
+	}
+	p.cache = make([]*cachedEntry, 0)
+	return nil
+}
+
+// PruneCache removes all entries from the cache at the given height
+func (p *provableStore) PruneCache(store kvstore.KVStore, height uint64) error {
+	keys, _, err := store.GetAll([]byte(fmt.Sprintf("%s/%d", p.name, height)), false)
```
Comment (Olshansk) [inline ibc/store/provable_store.go:200]: Have you thought about about:

1. Updating `PersistenceModule` to allowing returning the writeContext (it's a singleton initialized by NewRWContext)

2. Using the write context for the cache? That way as it errors or is released, the logic is centralized in one place.

cc @dylanlott for thoughts on this as well

File: ibc/store/provable_store.go [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+	return nil
+}
+
+// RestoreCache loads all entries from disk into the cache
+func (p *provableStore) RestoreCache(store kvstore.KVStore) error {
+	keys, values, err := store.GetAll(p.prefix, false)
+	if err != nil {
+		return err
+	}
+	for i := 0; i < len(keys); i++ {
```
Comment (Olshansk) [inline ibc/store/provable_store.go:218]: for i, key := range keys

File: ibc/module.go [hunk 19]
```diff
@@ -75,13 +86,73 @@ func (m *ibcModule) GetModuleName() string {
...
+		return err
+	}
+	signature, err := pk.Sign(signableBz)
+	if err != nil {
+		return err
+	}
+	tx.Signature = &coreTypes.Signature{
+		Signature: signature,
+		PublicKey: pk.PublicKey().Bytes(),
+	}
```
[the remark this answers is not in the document]
Comment (h5law) [reply ibc/module.go:133]: So this is addressed in the host submodule PR. It should not be this way but until the logic for emitting the messages was inplace they had to be signed upon receipt. I have actually removed this function altogether and now pass the signed tx straight to utility. See #866

File: ibc/store/provable_store.go (hunk 17 above)
Comment (h5law) [reply ibc/store/provable_store.go:200]: A little unsure of what you mean. Having a new cache table for example in the postgres DB and using that as our cache? The idea behind the cache is that if the store change event fails to propagate or the node crashes we can retry. I assumed an in memory cache which is periodically written to disk and pruned (handled by store manager soon TM) would be better but happy to discuss this further.

**Comment by h5law** [comment]
> Can you please merge with main before we take another look at this?

@Olshansk Dunno why it did this, I merged with main earlier today o-o did again should be fine now

File: ibc/module.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply ibc/module.go:133]: > So this is addressed in the host submodule PR

Cool, got it.

>  It should not be this way but until the logic for emitting the messages was inplace they had to be signed upon receipt

As a reviewer, I spent a lot of time reading, thinking and trying to understand it.

Totally fine (and encouraged) to have interim work but remember to always put yourself in the reviewer's shoes. Or, what if I had to pick this up myself tomorrow? I'd be wondering why it is this way. This is one (of many) reasons why cosmos ended up becoming the way it is.

For example, adding a comment similar to

```go
// TECHDEBT(#866): This SHOULD NOT be this way and is actually just a temporary workaround. The correct thing to do is ... and the code we'll be deleting is ...
```

File: persistence/trees/trees.go (hunk 16 above)
Comment (Olshansk) [reply persistence/trees/trees.go:373]: Noted, maybe just use an `else if` instead?

File: ibc/store/provable_store.go (hunk 17 above)
Comment (h5law) [reply ibc/store/provable_store.go:210]: added at top of this file

File: persistence/trees/trees.go (hunk 16 above)
Comment (h5law) [reply persistence/trees/trees.go:373]: changed to use a continue and removed 2nd branch

**Review by Olshansk — APPROVED** [review]
